attempt at improving the query guide (#22)
neo4j-graph-examples/get-started: documentation/query.adoc

== Get to know the database

A good place to start is to take stock of what you have in your database.

> You access _Database Information_ from the very top left of the sidebar.

Reviewer: The sidebar doesn't have anything lined up horizontally, only vertically, so I'm not sure if we need to specify _left_ here.

> You access _Database Information_ from the very top left of the sidebar.

Reviewer: ```suggestion
You access _Database Information_ from the top of the sidebar.
```

MATCH (n:Product) RETURN n LIMIT 25;
----

== Excecuting queries with graph results

Central to the Query UI is the Cypher editor on top where you write and run your queries and the list of result frames.

> When you _run_ any query a result frame shows which shows the query with its results, pushing previous ones down.

Reviewer: ```suggestion
When you _run_ any query, a result frame shows the query and its results, pushing previous ones down.
```

Try it out by typing in and executing the following Cypher query.

[source,cypher]
----
MATCH path=(:Product)-[:PART_OF]->(c:Category) 
WHERE c.categoryName = 'Produce'
RETURN path;
----

You can edit the query within the frame and run it again without creating a new result frame.

The frame has two important views: "Graph" and "Table". ("Plan" is for showing query plans and "RAW" for the raw results).

//highlight:query/result-view-graph[Result view]
//doesn't work

// image::result-options.png[]
//would be cool to replace this with a highlight also

The default view is graph if your query returns graph elements like nodes, relationships or paths.

NOTE: Relationships are only shown if you return a path, or name and return the relationships as well as the nodes.

> Here we see `Product` nodes with `PART_OF` relationships to the `Category` with name _Produce_.

Reviewer: ```suggestion
Here you can see `Product` nodes with `PART_OF` relationships to the `Category` with name _Produce_.
```

On the right side of the graph view you can see the properties panel with information about the entities in your graph.
You can click on any of them to change their styling, i.e. color, size, and caption, clicking on one element shows its attributes.

== Showing tabular results

If your query only returns scalar values (like strings or numbers), the result defaults to the table view and the graph view is not available.

Change your query to:

[source,cypher]
----
MATCH (p:Product)-[:PART_OF]->(c:Category) 
WHERE c.categoryName = 'Produce'
RETURN p.productName,c.categoryName;
----

> The query uses _variables_, `c` and `p`, for the category and the product as we want to refer to them later.

Reviewer: ```suggestion
The query uses _variables_, `c` and `p`, for the category and the product as you will want to refer to them later.
```

[source, cypher]
----
CALL db.schema.visualization()
----

Let's take a look at the different companies supplying Northwind with products and see which supplier provides products from which categories:

[source, cypher]
----
MATCH (s:Supplier)-->(:Product)-->(c:Category)
RETURN s.companyName as company, collect(distinct c.categoryName) as categories
----

This query finds _suppliers_ that _supply_ a _product_ which is _part of_ a _category_ and returns the company names and the categories they supply products from, in a table.
In cases where the same supplier supplies products from more than one category, the distinct category names are collected into a list, as you can see in the _categories_ column.

Even though you specified the _Product_ node in the `MATCH` part of the query, no information about the actual product is returned since this is not included in the `RETURN` part.

Maybe you are interested only in suppliers of a certain category, say _Condiments_:

[source, cypher]
----
MATCH (s:Supplier)-->(:Product)-->(c:Category)
WHERE c.categoryName = 'Condiments'
RETURN DISTINCT s.companyName as condimentsSuppliers
----

> Here we're not interested in the `Product` nodes, nor the exact types of the relationships between the nodes, so we don't have to specify them.

Reviewer: ```suggestion
Since you are not interested in the `Product` nodes, nor the exact types of the relationships between the nodes here, you don't have to specify them.
```

This is fairly basic, so let's move on to something a little more complex.

== Write more advanced Cypher for problem-solving

Assume that you want to see which product categories are typically co-ordered with other product categories and how frequently.

This might help you understand which products to promote alongside others.

[source, cypher]
----
// which categories are the products of an order in
MATCH (o:Order)-[:ORDERS]->(:Product)-[:PART_OF]->(c:Category)
// retain same ordering of categories
WITH o, c ORDER BY c.categoryName
// aggregate categories by order into a list of names
WITH o, collect(DISTINCT c.categoryName) as categories 
// only orders with more than one category
WHERE size(categories) > 1
// count how frequently the pairings occurr
RETURN categories, count(*) as freq
// order by frequency
ORDER BY freq DESC
LIMIT 50
----

> Another question would be, which customers are similar, i.e. ordering similar products most frequently.

Reviewer: ```suggestion
Another related question is which customers are similar, i.e. ordering similar products at the same frequency.
```

Author: That would change the meaning.

Reviewer: Hmm.. How about this:
_You can extend this and find customers who display similar purchasing patterns, i.e. which customers order similar products most frequently._

> The base question is the same, just that we expand across the product to other customers.

Reviewer: ```suggestion
The base question is the same, you just expand across the product to other customers.
```

> We find the "peer-groups" of our customers, which then can be used for product recommdations (people that bought X also bought) or segmentation into clusters of our customer base.

Reviewer: ```suggestion
You find the "peer-groups" of our customers, which then can be used for product recommendations (people who bought X also bought) or segmentation into clusters of your customer base.
```

[source, cypher]
----
// pattern from customer purchasing products to another customer purchasing the same products
MATCH (c:Customer)-[:PURCHASED]->(:Order)-[:ORDERS]->(p:Product)<-[:ORDERS]-(:Order)<-[:PURCHASED]-(c2:Customer)
// don't want the same customer pair twice
WHERE c < c2
// sort by the top-occuring products
WITH c, c2, p, count(*) as productOccurrence 
ORDER BY productOccurrence DESC
// return customer pairs ranked by similarity and the top 5 products
RETURN c.companyName, c2.companyName, sum(productOccurrence) as similarity, collect(distinct p.productName)[0..5] as topProducts
ORDER BY similarity DESC LIMIT 10
----

> Now we could create relationships for all customers that score more than 50 in our similarity score and see how they cluster.

Reviewer: ```suggestion
Now you could create relationships for all customers that score more than 50 in your similarity score and see how they cluster.
```
